Source/WebCore/editing/FrameSelection.cpp

```diff
@@ -1922,16 +1950,35 @@ void CaretBase::paintCaret(const Node& node, GraphicsContext& context, const Lay
     if (caret.isEmpty())
         return;
 
-    Color caretColor = Color::black;
+    Color oldCaretColor = Color::black;
     auto* element = is<Element>(node) ? downcast<Element>(&node) : node.parentElement();
     if (element && element->renderer())
-        caretColor = CaretBase::computeCaretColor(element->renderer()->style(), &node);
+        oldCaretColor = CaretBase::computeCaretColor(element->renderer()->style(), &node);
 
     auto pixelSnappedCaretRect = snapRectToDevicePixels(caret, node.document().deviceScaleFactor());
-    if (caretAnimator)
-        caretAnimator->paint(node, context, pixelSnappedCaretRect, caretColor, paintOffset, selection);
-    else
-        context.fillRect(pixelSnappedCaretRect, caretColor);
+    if (caretAnimator) {
+        auto caretColor = [&] {
+#if HAVE(REDESIGNED_TEXT_CURSOR)
+            if (!selection) {
+                auto* element = is<Element>(node) ? downcast<Element>(&node) : node.parentElement();
+                if (element && element->renderer())
+                    return platformCaretColor(element->renderer()->style(), &node);
+
+                return oldCaretColor;
+            }
+
+            if (RefPtr editableRoot = selection->rootEditableElement(); editableRoot && editableRoot->renderer())
+                return platformCaretColor(editableRoot->renderer()->style(), &node);
+#else
+            UNUSED_PARAM(selection);
+#endif
+
+            return oldCaretColor;
+        }();
```

Comment:
Shouldn't all this logic be in `CaretBase::computeCaretColor`? Seems strange to "compute" the color again in the painting method?

Reply:
yes; I planned on doing that refactor in a separate PR; didn't want to refactor any more than necessary in this one

Reply:
Why is this change even necessary as part of the upstreaming? I think there should be as few lines changed as possible during upstreaming to avoid potential regressions. Is it wrong to have:
```
if (caretAnimator)
        caretAnimator->paint(node, context, pixelSnappedCaretRect, caretColor, paintOffset, selection);
    else
        context.fillRect(pixelSnappedCaretRect, caretColor);
```
as we did before?

Reply:
@mwyrzykowski I originally had this, but ended up changing it because `platformCaretColor` was in one of the caret animators and it compiled fine previously, but would not compile after upstreaming the animator files, so I had to put it elsewhere and FrameSelection was a logical place

Reply:
Wouldn't as a static member function on the base class `CaretAnimator` be more appropriate since `platformCaretColor` is being using by `OpacityCaretAnimator` and `DictationCaretAnimator`?

Reply:
yes, that makes more sense. will change for this PR, and then combine platformcaretcolor with computecaretcolor in a follow up PR.